consensus/tendermint/tmdb.go

```diff
@@ -0,0 +1,311 @@
+		}
+	}
+	return walMsgs, nil
+}
+
+// scanWALRaw iterates over raw WAL entries in the database for a given height.
+func (s *tendermintDB[V, H, A]) scanWALRaw(height height) ([][]byte, error) {
+	startKey := tmdb.WALEntry.Key(encodeNumMsgsAtHeight(walIter(height)))
+	rawEntries := [][]byte{}
+
+	err := s.db.View(func(snap db.Snapshot) error {
+		iter, err := snap.NewIterator(startKey, false)
+		if err != nil {
+			return err
+		}
+		defer iter.Close()
+
+		if !iter.Seek(startKey) {
+			// Changed error message slightly for clarity
+			return fmt.Errorf("failed to seek to start key when scanning WAL msgs")
+		}
+		if !iter.Valid() {
+			// It's possible there are no entries, return empty slice and no error
+			return nil
+		}
+
+		for ; iter.Valid(); iter.Next() {
+			// Check if the key still has the correct prefix (height)
+			key := iter.Key()
+			if !bytes.HasPrefix(key, startKey) {
+				break // Stop if we moved past keys for this height
+			}
+			v, err := iter.Value()
+			if err != nil {
+				return fmt.Errorf("scanWALRaw: failed to get value: %w", err)
+			}
+			valueCopy := make([]byte, len(v))
+			copy(valueCopy, v)
+			rawEntries = append(rawEntries, valueCopy)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("scanWALRaw: db view error: %w", err)
+	}
+	return rawEntries, nil
+}
+
+// decodeWALMessageData decodes the inner CBOR message data based on the provided type string.
+// It assumes the outer wrapper has already been decoded.
+func decodeWALMessageData[V Hashable[H], H Hash, A Addr](msgType MessageType, data cbor.RawMessage) (walEntry, error) {
+	switch msgType {
+	case MessageTypeProposal:
+		var proposal Proposal[V, H, A]
+		if err := cbor.Unmarshal(data, &proposal); err != nil {
+			return walEntry{}, fmt.Errorf("decodeWALMessageData: Proposal unmarshal failed: %w", err)
+		}
+		return walEntry{Type: MessageTypeProposal, Entry: proposal}, nil
+	case MessageTypePrevote:
+		var vote Prevote[H, A]
+		if err := cbor.Unmarshal(data, &vote); err != nil {
+			return walEntry{}, fmt.Errorf("decodeWALMessageData: Prevote unmarshal failed: %w", err)
+		}
+		return walEntry{Type: MessageTypePrevote, Entry: vote}, nil
+	case MessageTypePrecommit:
+		var vote Precommit[H, A]
+		if err := cbor.Unmarshal(data, &vote); err != nil {
+			return walEntry{}, fmt.Errorf("decodeWALMessageData: Precommit unmarshal failed: %w", err)
+		}
+		return walEntry{Type: MessageTypePrecommit, Entry: vote}, nil
+	default:
+		return walEntry{}, fmt.Errorf("decodeWALMessageData: unknown message type string %q", msgType.String())
+	}
+}
+
+func heightToBytes(height height) []byte {
+	heightBytes := make([]byte, NumBytesForHeight)
+	binary.BigEndian.PutUint32(heightBytes, uint32(height))
+	return heightBytes
+}
+
+func encodeNumMsgsAtHeight(numMsgsAtHeight walIter) []byte {
+	numMsgsAtHeightBytes := make([]byte, NumBytesForHeight)
+	binary.BigEndian.PutUint32(numMsgsAtHeightBytes, uint32(numMsgsAtHeight))
+	return numMsgsAtHeightBytes
+}
```

Comment:
Personally I think that we should have `Marshaler` and `Unmarshaler inferfaces for DB keys and the functions should be `MarshalKey` and `UnmarshalKey`. Not in this PR though.
For now, I think let's just have a consistent naming between these 2 functions.